homeassistant/components/rointe/device_manager.py

```diff
@@ -0,0 +1,384 @@
+                    )
+
+                    devices[device_id].hass_available = False
+
+        return discovered_devices
+
+    async def _get_firmware_map(self) -> dict[str, str] | None:
+        """Retrieve the latest firmware map for all rointe products."""
+        firmware_map_response: ApiResponse = await self.hass.async_add_executor_job(
+            self.rointe_api.get_latest_firmware
+        )
+
+        if firmware_map_response and firmware_map_response.success:
+            return firmware_map_response.data
+
+        LOGGER.error(
+            "Unable to fetch Rointe firmware update map: %s",
+            firmware_map_response.error_message,
+        )
+        return None
+
+    def _add_or_update_device(
+        self,
+        device_data,
+        energy_stats: EnergyConsumptionData,
+        device_id: str,
+        latest_fw: str | None,
+    ) -> list[RointeDevice] | None:
+        """Process a device from the API and add or update it.
+
+        Return a list of newly discovered devices.
+        """
+
+        device_data_data = device_data.get("data", None)
+        new_devices = []
+
+        if not device_data_data:
+            LOGGER.error("Device ID %s has no valid data. Ignoring", device_id)
+            return None
+
+        if device_id in self.rointe_devices:
+            # Existing device, update it.
+            target_device = self.rointe_devices[device_id]
+
+            if not target_device.hass_available:
+                LOGGER.debug("Restoring device %s", target_device.name)
+
+            target_device.update_data(device_data, energy_stats, latest_fw)
+
+            LOGGER.debug(
+                "Updating [%s] => Power: %s, Status: %s, Mode: %s, Temp: %s",
+                device_data_data.get("name", "N/A"),
+                target_device.power,
+                target_device.preset,
+                target_device.mode,
+                target_device.temp,
+            )
+        else:
+            # New device.
+
+            try:
+                device_type = device_data["data"]["type"]
+
+                if device_type not in ROINTE_SUPPORTED_DEVICES:
+                    LOGGER.warning("Ignoring Rointe device of type %s", device_type)
+                    return None
+
+                LOGGER.debug(
+                    "Found new device %s [%s] - %s",
+                    device_data_data.get("name", "N/A"),
+                    device_data_data.get("type", "N/A"),
+                    device_data_data.get("product_version", "N/A"),
+                )
+
+                rointe_device = RointeDevice(
+                    device_info=device_data,
+                    device_id=device_id,
+                    energy_data=energy_stats,
+                    latest_fw=latest_fw,
+                )
+
+                new_devices.append(rointe_device)
+                self.rointe_devices[device_id] = rointe_device
+
+            except Exception as ex:  # pylint: disable=broad-except
```

Comment:
Can we have a more specific exception here? It's generally bad form to catch broad exceptions unless it's a "I've caught every other type; I really don't know what this is."

Additionally, it isn't clear which of the elements of this try/except can actually generate the expression. I would suggest more, small try/except blocks to make it very clear.

Reply:
Actually there wasn't anything that explicitly throws exceptions and I ended removing the try-except block.